sandbox/apparmor/apparmor.go

```diff
@@ -56,6 +57,221 @@ func ValidateNoAppArmorRegexp(s string) error {
+		}
+	}
+
+	// Find the longest prefix common to ALL patterns.
+	commonPrefix := ""
+findCommonPrefix:
+	for charInd := 0; charInd < size; charInd++ {
+		for _, pattern := range excludePatterns {
+			if (charInd >= len(pattern)) || (pattern[charInd] != excludePatterns[0][charInd]) {
+				break findCommonPrefix
+			}
+		}
+		commonPrefix = excludePatterns[0][:charInd+1]
+	}
+
+	// This loop will iterate over the length of the longest excludePattern
+	// (charInd = 1..size), generating an apparmor rule on each iteration
+	// for the corresponding subpatterns, understanding as such, the first
+	// (charInd+1) characters of the excludePatterns.
+	builder := &strings.Builder{}
+	for charInd := 1; charInd < size; charInd++ {
+
+		// This loop will group the subpatterns properly, generating the subpatterns map, where:
+		//     - the key would be the subpatternPrefix, considering as such the subpattern except
+		//       its last character (pattern[0:charInd]).
+		//     - the value would be the charset, which would be the subpattern last character
+		//       (pattern[charInd]). If several subpatterns share the same subpatternPrefix, the
+		//       charset would be a string including the last character of all those subpatterns.
+		subpatternPrefix := ""
+		subpatterns := map[string]string{}
+		for _, pattern := range excludePatterns {
+			// Handle unsupported cases
+			if (charInd < len(pattern)) && pattern[charInd] == '*' {
+				// Check if the excludePattern has a character different from '/' after a wildcard
+				if ((charInd + 1) < len(pattern)) && (pattern[charInd+1] != '/') {
+					return "", errors.New("exclude patterns does not support suffixes for now")
+				}
+				// Check if '*' is the first character after a common subpattern
+				for _, patt := range excludePatterns {
+					if (patt != pattern) && ((charInd) < len(patt)) && (pattern[:charInd] == patt[:charInd]) && (patt[charInd] != '*') {
+						return "", errors.New("first character after a common subpattern cannot be a wildcard")
+					}
+				}
+			}
+
+			// Skip patterns that are already finished, wildcards and slashes preceded by wildcards.
+			if (charInd >= len(pattern)) ||
+				(pattern[charInd] == '*') ||
+				((pattern[charInd] == '/') && (pattern[charInd-1] == '*')) {
+				continue
+			}
+			// Group subpatterns
+			subpatternPrefix = pattern[:charInd]
+			if charset, exists := subpatterns[subpatternPrefix]; !exists {
+				// Add the pattern if it didn't exist yet
+				subpatterns[subpatternPrefix] = string(pattern[charInd])
+			} else {
+				if !strings.Contains(charset, string(pattern[charInd])) {
+					// Two patterns only differ on the last character
+					subpatterns[subpatternPrefix] = charset + string(pattern[charInd])
+				}
+			}
+		}
+
+		// Write patterns
+		if len(subpatterns) == 0 {
+			continue
+		} else if len(subpatterns) == 1 {
+			res := fmt.Sprintf("%s%s[^%s]**%s\n", opts.Prefix, subpatternPrefix, subpatterns[subpatternPrefix], opts.Suffix)
```

Comment:
instead of checking for len == 1 here, we could share more code and check for len(aux) below, as anyway len == 1 is not the common case?

Reply:
It will be 1, at least for all the iterations of the outer loop `for charInd := 1; charInd < size; charInd++` where `charInd < len(commonPrefix)`. 

As in the previous comment, we could move the apparmor rules generation for the common prefix out of the main loop (for charInd := 1; charInd < size; charInd++ ), however it will require to duplicate some checks (suffixes, '' and '/').